IndexedDB/name-scopes.html

```diff
@@ -0,0 +1,151 @@
+<!DOCTYPE html>
+<meta charset="utf-8">
+<title>
+  IndexedDB: scoping for database / object store / index names, and index keys
+</title>
+<link rel="help" href="https://w3c.github.io/IndexedDB/#constructs">
+<link rel="author" href="pwnall@chromium.org" title="Victor Costan">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="support-promises.js"></script>
+<script>
+
+// Creates the structure inside a test database.
+//
+// The structure includes two stores with identical indexes and nearly-similar
+// records. The records differ in the "path" attribute values, which are used to
+// verify that IndexedDB returns the correct records when queried.
+//
+// databaseName appears redundant, but we don't want to rely on database.name.
+const buildStores = (database, databaseName, useUniqueKeys) => {
+  for (let storeName of ['x', 'y']) {
+    const store = database.createObjectStore(
+        storeName, { keyPath: 'pKey', autoIncrement: true });
+    for (let indexName of ['x', 'y']) {
+      store.createIndex(
+          indexName, `${indexName}Key`, { unique: useUniqueKeys });
+    }
+
+    for (xKeyRoot of ['x', 'y']) {
+      for (yKeyRoot of ['x', 'y']) {
+        let xKey, yKey;
+        if (useUniqueKeys) {
+          xKey = `${xKeyRoot}${yKeyRoot}`;
+          yKey = `${yKeyRoot}${xKeyRoot}`;
+        } else {
+          xKey = xKeyRoot;
+          yKey = yKeyRoot;
+        }
+        const path = `${databaseName}-${storeName}-${xKeyRoot}-${yKeyRoot}`;
+        store.put({ xKey: xKey, yKey: yKey, path: path });
+      }
+    }
+  }
+};
+
+// Creates two databases with identical structures.
+const buildDatabases = (testCase, useUniqueKeys) => {
+  return createNamedDatabase(testCase, 'x', database => {
+    return buildStores(database, 'x', useUniqueKeys);
+  }).then(database => {
+    return database.close();
+  }).then(() => createNamedDatabase(testCase, 'y', database => {
+    return buildStores(database, 'y', useUniqueKeys);
+  })).then(database => {
+    return database.close();
+  });
+};
+
+// Reads all the store's values using an index.
+//
+// Returns a Promise that resolves with an array of values.
+const readIndex = (testCase, index) => {
+  let resolve = null;
+  const promise = new Promise((_resolve, _reject) => resolve = _resolve);
```

Comment:
Why this pattern (capturing `resolve`) rather than the more idomatic style of putting everything in the body of the Promise callback?